src/transformers/modeling_utils.py

```diff
@@ -353,6 +353,13 @@ def check_support_param_buffer_assignment(model_to_load, state_dict, start_prefi
     if is_deepspeed_zero3_enabled():
         return False
 
+    try:
+        from deepspeed.moe.utils import has_moe_layers
+        if has_moe_layers(model_to_load)[0]:
+            return False
+    except:
+        pass
+
```

Comment:
Does this happen with all deepspeed models ?

Reply:
Yes, This affects all models using DeepSpeed MoE.

I have provided a simple demo to reproduce the problem. I believe the key issue is that DeepSpeed adds new attributes (`allreduce` and `group_name`) to the parameters in `deepspeed.moe.experts.Experts.__init__`.

```python
from transformers import AutoConfig, AutoModelForCausalLM, AutoProcessor, PreTrainedModel, PretrainedConfig
import torch
from torch import nn
import deepspeed

class MoEConfig(PretrainedConfig):
    model_type = "simple_moe"

    def __init__(
            self,
            hidden_size=4096,
            num_experts=4,
            k=2,
            ep_size=1,
            pad_token_id=None,
            bos_token_id=1,
            eos_token_id=2,
            tie_word_embeddings=False,
            **kwargs,
    ):
        self.hidden_size = hidden_size
        self.num_experts = num_experts
        self.k = k
        self.ep_size = ep_size

        super().__init__(
            pad_token_id=pad_token_id,
            bos_token_id=bos_token_id,
            eos_token_id=eos_token_id,
            tie_word_embeddings=tie_word_embeddings,
            **kwargs,
        )


class MoE(PreTrainedModel):
    config_class = MoEConfig

    def _init_weights(self, module):
        pass

    def __init__(self, config):
        super().__init__(config)
        self.moe = deepspeed.moe.layer.MoE(
            hidden_size=config.hidden_size,
            expert=nn.Linear(config.hidden_size,config.hidden_size),
            num_experts=config.num_experts,
            ep_size=config.ep_size,
            k=config.k
        )

    def forward(self, x):
        x, l_aux, exp_counts = self.moe(x)
        return x

AutoConfig.register("simple_moe", MoEConfig)
AutoModelForCausalLM.register(MoEConfig, MoE)

config = {
    "hidden_size": 256,
    "num_experts": 4,
    "ep_size": 1,
    "k": 2
}
config = MoEConfig.from_dict(config)
model = MoE(config)

from deepspeed.moe.utils import is_moe_param
# Using the original model, it can be recognized as MoE
for n,p in model.named_parameters():
    print("check-original", n, is_moe_param(p))

# Loading the model from local storage with the same torch dtype,
# it cannot be recognized as MoE. This will throw an exception
# during the MoE parameter check in DeepSpeed
# (deepspeed.runtime.zero.stage_1_and_2.DeepSpeedZeroOptimizer._configure_moe_settings)
model.save_pretrained("./tmp/")
model = AutoModelForCausalLM.from_pretrained("./tmp/", torch_dtype=model.dtype)
for n,p in model.named_parameters():
    print("check-same_type", n, is_moe_param(p))

# Loading the model from local storage with a different torch dtype,
# it can be recognized as MoE
model = AutoModelForCausalLM.from_pretrained("./tmp/", torch_dtype=torch.bfloat16)
for n,p in model.named_parameters():
    print("check-diff_dtype", n, is_moe_param(p))

```